<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate mosaic (multi-batch) single-cell MTX data for eQTL analysis. — make.sc.eqtl.mosaic • mmutilR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate mosaic (multi-batch) single-cell MTX data for eQTL analysis. — make.sc.eqtl.mosaic"><meta property="og:description" content="We will generate Y ~ Poisson(mu * rho) where mu ~ exp(log.mu/smudge), rho ~ Gamma(a,b)"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mmutilR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate mosaic (multi-batch) single-cell MTX data for eQTL analysis.</h1>
    
    <div class="hidden name"><code>make.sc.eqtl.mosaic.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>We will generate Y ~ Poisson(mu * rho) where mu ~ exp(log.mu/smudge), rho ~ Gamma(a,b)</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">make.sc.eqtl.mosaic</span><span class="op">(</span></span>
<span>  <span class="va">file.header</span>,</span>
<span>  <span class="va">X</span>,</span>
<span>  <span class="va">h2</span>,</span>
<span>  n.causal.snps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  n.causal.genes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  pve.y.by.u0 <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  n.u0 <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  pve.u1.by.x <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  pve.y.by.u1 <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  n.u1 <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  pve.interaction <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  n.interaction <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  n.genes <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  n.covar.genes <span class="op">=</span> <span class="va">n.genes</span>,</span>
<span>  num.mixtures <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  num.mosaic <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  smudge <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  rho.a <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  rho.b <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ncell.ind <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  rseed <span class="op">=</span> <span class="fl">13</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>X</dt>
<dd><p>genotype matrix (individual x SNPs)</p></dd>


<dt>h2</dt>
<dd><p>heritability (proportion of variance of Y explained by genetic X)</p></dd>


<dt>n.causal.snps</dt>
<dd><p>X variables directly affecting on Y</p></dd>


<dt>n.causal.genes</dt>
<dd><p>Y variables directly regulated by X</p></dd>


<dt>pve.y.by.u0</dt>
<dd><p>proportion of variance of Y explained by U0</p></dd>


<dt>n.u0</dt>
<dd><p>number of covariates on Y</p></dd>


<dt>pve.u1.by.x</dt>
<dd><p>proportion of variance of U1 explained by X</p></dd>


<dt>pve.y.by.u1</dt>
<dd><p>proportion of variance of Y explained by U1</p></dd>


<dt>n.u1</dt>
<dd><p>number of covariates on Y</p></dd>


<dt>pve.interaction</dt>
<dd><p>proportion of variance of Y explained by interaction</p></dd>


<dt>n.interaction</dt>
<dd><p>number of genes interacting with the causal genes</p></dd>


<dt>n.genes</dt>
<dd><p>total number of genes (Y variables)</p></dd>


<dt>num.mixtures</dt>
<dd><p>num of cell mixtures</p></dd>


<dt>smudge</dt>
<dd><p>a scaling factor for a GLM model (default: 1)</p></dd>


<dt>rho.a</dt>
<dd><p>rho ~ Gamma(a, b)</p></dd>


<dt>rho.b</dt>
<dd><p>rho ~ Gamma(a, b)</p></dd>


<dt>ncell.ind</dt>
<dd><p>number of cells per individual</p></dd>


<dt>rseed</dt>
<dd><p>random seed</p></dd>


<dt>num.batches</dt>
<dd><p>num of single-cell data batches</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>simulation results</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The simulation result list will have two lists:</p>
<p><code>data</code>:</p><ul><li><p><code>data$mtx</code>: a matrix market data file</p></li>
<li><p><code>data$row</code>: a file with row names</p></li>
<li><p><code>data$col</code>: a file with column names</p></li>
<li><p><code>data$idx</code>: an indexing file for the columns</p></li>
<li><p><code>data$indv</code>: a mapping file between column and individual names</p></li>
</ul><p><code>indv</code>:</p><ul><li><p><code>indv$y</code>: observed (noisy) individual x gene matrix</p></li>
<li><p><code>indv$x</code>: observed individual x variants genotype matrix</p></li>
<li><p><code>indv$causal.snps</code>: causal variants (X variables)</p></li>
<li><p><code>indv$causal.genes</code>: causal genes (Y variables)</p></li>
<li><p><code>indv$causal.label</code>: true labels</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yongjin Park.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

